embassy-stm32/src/dma/ringbuffer.rs

```diff
@@ -54,151 +53,130 @@ pub trait DmaCtrl {
-        let end = self.end();
+    pub fn read(&mut self, mut dma: impl DmaCtrl, buf: &mut [W]) -> Result<(usize, usize), OverrunError> {
+        /*
+            This algorithm is optimistic: we assume we haven't overrun more than a full buffer and then check
+            after we've done our work to see we have. This is because on stm32, an interrupt is not guaranteed
+            to fire in the same clock cycle that a register is read, so checking get_complete_count early does
+            not yield relevant information.
+
+            Therefore, the only variable we really need to know is ndtr. If the dma has overrun by more than a full
+            buffer, we will do a bit more work than we have to, but algorithms should not be optimized for error
+            conditions.
+
+            After we've done our work, we confirm that we haven't overrun more than a full buffer, and also that
+            the dma has not overrun within the data we could have copied. We check the data we could have copied
+            rather than the data we actually copied because it costs nothing and confirms an error condition
+            earlier.
+        */
+        let end = self.pos(dma.get_remaining_transfers());
+        if self.start == end && dma.get_complete_count() == 0 {
+            // No bytes are available in the buffer
+            Ok((0, self.len()))
+        } else if self.start < end {
+            // The available, unread portion in the ring buffer DOES NOT wrap
+            // Copy out the bytes from the dma buffer
+            let len = self.copy_to(buf, self.start..end);
 
-        compiler_fence(Ordering::SeqCst);
+            compiler_fence(Ordering::SeqCst);
 
-        if self.first == end {
-            // The buffer is currently empty
+            /*
+                first, check if the dma has wrapped at all if it's after end
+                or more than once if it's before start
 
-            if dma.get_complete_count() > 0 {
-                // The DMA has written such that the ring buffer wraps at least once
-                self.ndtr = dma.ndtr();
-                if self.end() > self.first || dma.get_complete_count() > 1 {
-                    return Err(OverrunError);
-                }
-            }
+                this is in a critical section to try to reduce mushy behavior.
+                it's not ideal but it's the best we can do
 
-            Ok(0)
-        } else if self.first < end {
-            // The available, unread portion in the ring buffer DOES NOT wrap
+                then, get the current position of of the dma write and check
+                if it's inside data we could have copied
+            */
+            let (pos, complete_count) =
+                critical_section::with(|_| (self.pos(dma.get_remaining_transfers()), dma.get_complete_count()));
+            if (complete_count > 0 && pos >= end) || complete_count > 1 || (pos >= self.start && pos < end) {
```

Comment:
Do we need some `compiler_fence(Ordering::SeqCst)` to guarantee the ordering here?

Reply:
> the dma is currently writing before the block that we have copied out.

We don't know that. Assuming we still had valid data, the dma could still be writing after the end or it could have wrapped. If it has wrapped (`complete_count > 0` guarantees it has wrapped at least once but we cannot make any other claim), then we know for certain than if the position is after the end of our buffer, the dma has overwritten data that could have been valid.

Reconciling this logical condition with that condition, this condition can detect a case where `complete_count` has not yet 
 been updated but the dma has overwritten the data. In other words, it has a greater tolerance for interrupt lag. Whether this would actually happen is debatable but it is still correct.
 
 >  I dont think the cs is needed, but I may be wrong.
 
 It is needed because I could not forsee the impact that a long interrupt could have if it occurred between the assignment of pos and complete count.
 
 > Do we need some compiler_fence(Ordering::SeqCst) to guarantee the ordering here?
 
 You may be right about this. I have no objection to this change.

Reply:
Year, maybe I just have to give up in understanding this. I was specifically requested to take out the code that handled interrupt tolerance in `get_complete_count()` which was previously [there](https://github.com/rmja/embassy/blob/45843034ec7c53f3f619bf1dcc6c145c428b0601/embassy-stm32/src/dma/dma.rs#L636-L643) in the PR - it tested the tcif flag, and adjusted complete_count accordingly, not relying on the dma to actually have run its isr.